### PR TITLE
Add dependency on go.uber.org/automaxprocs

### PR DESCRIPTION
**Pull request by TomasKohout**
Signed-off-by: Tomas Kohout <tomas.kohout1995@gmail.com>

Hi, 

this PR is about adding a new dependency on go.uber.go/automaxprocs. 

Why we need that?

As you all know nodes come with different sizes. For example, we use quite huge nodes with up to 104 CPUs and because we use prometheus-operator, we aren't able to set environment variables to set GOMAXPROCS to the value of `cpu.limit`. automaxprocs package does this automatically without any additional pain. If there are not limits specified, default value (number of cores) will be left in place.

### CODE REVIEW

**Comment by roidelapluie** [comment]
Thanks for your contribution. Until the go runtime figures this out, can we have this behind a feature flag rather than having it enabled by default?

**Comment by bboreham** [comment]
I think having a side-effect caused by `_` is a bit too subtle.
I expect adding a feature flag will fix this too - there will be some description associated with the name.

**Comment by peterbourgon** [comment]
I agree with @roidelapluie and @bboreham.

> because we use prometheus-operator, we aren't able to set environment variables

Is this a limitation of prometheus-operator specifically, or operators in general?

**Comment by TomasKohout** [comment]
Feature flag added. :slightly_smiling_face:

**Comment by TomasKohout** [comment]
> I agree with @roidelapluie and @bboreham.
> 
> > because we use prometheus-operator, we aren't able to set environment variables
> 
> Is this a limitation of prometheus-operator specifically, or operators in general?

It depends. Maybe there are some operators which expose env field via theirs CRD. Prometheus-operator doesn't do so.

**Comment by TomasKohout** [comment]
I'll add feature description to the documentation tomorrow.

File: cmd/prometheus/main.go [hunk 1]
```diff
@@ -186,6 +189,9 @@ func (c *flagConfig) setFeatureListOptions(logger log.Logger) error {
 			case "promql-per-step-stats":
 				c.enablePerStepStats = true
 				level.Info(logger).Log("msg", "Experimental per-step statistics reporting")
+			case "auto-gomaxprocs":
+				c.enableAutoGOMAXPROCS = true
+				level.Info(logger).Log("msg", "Automatically set GOMAXPROCS to cpu.limits when running in container.")
```
Comment (peterbourgon) [inline cmd/prometheus/main.go:194]:
```suggestion
				level.Info(logger).Log("msg", "Automatically set GOMAXPROCS to cpu.limits when running in container")
```

File: cmd/prometheus/main.go [hunk 2]
```diff
@@ -564,6 +570,14 @@ func main() {
 		ruleManager *rules.Manager
 	)
 
+	if cfg.enableAutoGOMAXPROCS {
+		_, err := maxprocs.Set()
+
+		if err != nil {
+			level.Warn(logger).Log("failed to set GOMAXPROCS: %v", err)
+		}
```
Comment (peterbourgon) [inline cmd/prometheus/main.go:578]:
```suggestion
		if _, err := maxprocs.Set(); err != nil {
			level.Warn(logger).Log("auto_gomaxprocs", "failed", "err", err)
		}
```

File: cmd/prometheus/main.go [hunk 3]
```diff
@@ -32,6 +32,8 @@ import (
 	"syscall"
 	"time"
 
+	"go.uber.org/automaxprocs/maxprocs"
+
```
Comment (peterbourgon) [inline cmd/prometheus/main.go:36]: Probably should group with other third-party imports.

File: cmd/prometheus/main.go (hunk 1 above)
Comment (peterbourgon) [inline cmd/prometheus/main.go:194]: If this feature flag is enabled, but Prometheus is not running in a container, will the `maxprocs.Set` call (below) fail and emit a warning?

File: cmd/prometheus/main.go [hunk 4]
```diff
@@ -564,6 +570,14 @@ func main() {
 		ruleManager *rules.Manager
 	)
 
+	if cfg.enableAutoGOMAXPROCS {
+		_, err := maxprocs.Set()
+
+		if err != nil {
+			level.Warn(logger).Log("failed to set GOMAXPROCS: %v", err)
```
Comment (roidelapluie) [inline cmd/prometheus/main.go:577]:
```suggestion
			level.Warn(logger).Log("msg", "Failed to set GOMAXPROCS automatically", "err", err)
```

File: cmd/prometheus/main.go (hunk 1 above)
Comment (roidelapluie) [inline cmd/prometheus/main.go:194]:
```suggestion
				level.Info(logger).Log("msg", "Automatically set GOMAXPROCS to match Linux container CPU quota")
```

Comment (TomasKohout) [reply cmd/prometheus/main.go]: It'll not fail not emit warning, it'll leave the default which is equal to number of CPUs.

I'm not sure under which conditions would Set fail. I've added warning for user to know why the GOMAXPROCS is not set when this feature is enabled.

File: cmd/prometheus/main.go [hunk 5]
```diff
@@ -564,6 +569,12 @@ func main() {
 		ruleManager *rules.Manager
 	)
 
+	if cfg.enableAutoGOMAXPROCS {
+		if _, err := maxprocs.Set(); err != nil {
+			level.Warn(logger).Log("auto_gomaxprocs", "failed", "msg", "Failed to set GOMAXPROCS automatically", "err", err)
```
Comment (roidelapluie) [inline cmd/prometheus/main.go:574]:
```suggestion
			level.Warn(logger).Log("msg", "Failed to set GOMAXPROCS automatically", "err", err)
```

to match what we have everywhere else in the codebase.

**Comment by roidelapluie** [comment]
This LGTM besides my comment. Can you please add a note in the Feature Flags documentation?
Thanks!

File: cmd/prometheus/main.go (hunk 5 above)
Comment (TomasKohout) [reply cmd/prometheus/main.go:574]: Ok, sorry.

I've merged suggestions from you and Peter.

**Comment by TomasKohout** [comment]
Note added. I've tried to describe it, but I'm not sure that it's described good enough. :slightly_smiling_face:  Could you have an additional look please?

**Review by bwplotka — APPROVED** [review]
Great, thanks! 💪

**Comment by TomasKohout** [comment]
Yay, sure thing!

**Comment by roidelapluie** [comment]
Thanks!

**Comment by TomasKohout** [comment]
That was fast! I should thank you for help. :slightly_smiling_face:

**Comment by SuperQ** [comment]
@TomasKohout I'm not sure where you get the idea that the Prometheus-Operator can't inject env vars.

From [the CRD API docs](https://github.com/prometheus-operator/prometheus-operator/blob/main/Documentation/api.md#prometheusspec)

`containers`:
> Containers allows injecting additional containers or modifying operator generated containers. This can be used to allow adding an authentication proxy to a Prometheus pod or to change the behavior of an operator generated container. Containers described here modify an operator generated container if they share the same name and modifications are done via a strategic merge patch. The current container names are: prometheus, config-reloader, and thanos-sidecar. Overriding containers is entirely outside the scope of what the maintainers will support and by doing so, you accept that this behaviour may break at any time without notice.

You should easily be able to something like this

```yaml
containers:
- name: prometheus
  env:
  - name: GOMAXPROCS
    value: 16
```

**Comment by TomasKohout** [comment]
@SuperQ well, that sound nice, but why should I do something that could break at any time without notice?

**Comment by bboreham** [comment]
May I suggest editing the PR title and description to match what was merged?
E.g. "Add auto-gomaxprocs optional feature" for the title.

**Comment by bboreham** [comment]
No helm chart in this repo.  Probably best to ask wherever the one you use is maintained.

**Pull request merged**